### DISCUSSION TITLE
Command prompt - missed new line and wrongly appended to program output

### DISCUSSION THREAD

**Discussion by Torinde** [dosbox-staging/dosbox-staging#3173] [General]
For a long time I felt Staging command prompt lines are "compressed". Now I made screenshots:

Staging:
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/dd063cd6-8a24-46e0-9bce-d467141593df)

X:
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/38259818-c99a-4729-9bd3-0b1ee7685da8)

VirtualBox MS-DOS 6.22:
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/7826d71b-ba7f-4360-91de-e19eed966b5c)

It seems Staging applies one less new line than the others (leading to having the prompt squashed next to the error message "Program abnormally terminated" here). Maybe it's a conscious decision to save space, but notifying in case developers consider it a bug.

**Comment by johnnovak** [comment]
Thanks @Torinde. I don't think it's conscious, probably just happened by fluke. Yeah I agree it looks better to have blank lines after the errors.

On the other hand, 3rd party programs always just did whatever. But I'm not against inserting an extra blank line after the output of all built-in commands, especially if that's what MS-DOS 6.22 does.

Btw, that's a nice custom font on the first screenshot. What is it and how did you set it?

**Reply by Torinde** [reply]
Blank line after DOS command errors is one thing (and more debatable), but what for sure looks off is the:
"Program abnormally terminated.C:\>" As if somebody changed my prompt=$P$G to contain that error...
I think both happen for the same reason - less blank lines that real DOS.

The font is coming from the codepage and keyboard layout TSR manager itself (BDSVGA) - for dual-alphabet users it's quite handy to switch between the two codepages on-the-fly (in this case Latin + Cyrillic) to see UI like boxes/menus written in 437. This TSR uses LeftShift+RightShift for that, hopefully Staging can implement that in KEYB.COM (and later in the OSD) - as mentioned [here](https://github.com/dosbox-staging/dosbox-staging/pull/3164#issuecomment-1837540852).

**Reply by Torinde** [reply]
FreeDOS appears to have "less" new lines similarly to Staging:
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/9408256b-e9b1-418b-95e9-626134be6a57)

So, maybe Staging mimicked FreeDOS instead of MS-DOS.

I'll rename this discussion accordingly.

**Reply by Torinde** [reply]
most characters from the font: 
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/3bc5ac7d-912a-45fc-9038-b8d99fe26a9a)
Above is from running it in Staging without shader or integer scalling and in the smallest window possible.

I would like having that as an option, so if a developer has interest in recreating it I can try to get the rest via Hex editor.

**Comment by Torinde** [comment]
After starting the BDSVGA TSR - both reverse-T arrow keys and Numpad keys give number characters if NumLock is ON. Same (wrong) behavior happens on X. VirtualBox with MS-DOS 6.22 works properly.

Should I open an Issue for that?

**Comment by Torinde** [comment]
The prompt appears properly on the next line - if you pipe the output trough MORE:
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/6cb9ebd1-10ef-4d0e-a47d-2942f7c0e38d)

I assume that's due to how MORE handles it.

**Comment by johnnovak** [comment]
Just tested booting into real MS-DOS 6.22 off a HDD image in Staging. All commands seem to insert a new line after their output.

And yeah, it looks a lot better, so we should normalise this behaviour in our command implementations.

<img width="1215" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/feb1e72f-b6a8-40b2-8591-911a9d6e2d58">

<img width="1215" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/1c8d8c15-1506-4f52-8b4c-664943a20531">

<img width="1215" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/64403e52-0bb0-4ea8-8afa-ae1988a50498">

**Comment by johnnovak** [comment]
I have a WIP PR for this, stay tuned.

**Reply by Torinde** [reply]
Did you merge that PR?

With current main I still get the old behavior of missing lines:
<img width="765" height="586" alt="image" src="https://github.com/user-attachments/assets/f9b1972a-1f8c-45ef-af2c-bbc7e179a734" />

**Reply by johnnovak** [reply]
Postponed to later this year.
